zen3geo/datapipes/pystac_client.py

```diff
@@ -133,3 +133,101 @@ def __iter__(self) -> Iterator:
+        JSON format.
+
+    Raises
+    ------
+    ModuleNotFoundError
+        If ``pystac_client`` is not installed. See
+        :doc:`install instructions for pystac-client <pystac_client:index>`,
+        (e.g. via ``pip install pystac-client``) before using this class.
+
+    Example
+    -------
+    >>> import pytest
+    >>> pystac_client = pytest.importorskip("pystac_client")
+    ...
+    >>> from torchdata.datapipes.iter import IterableWrapper
+    >>> from zen3geo.datapipes import PySTACAPIItemLister
+    ...
+    >>> # List STAC Items from a STAC API query
+    >>> catalog = pystac_client.Client.open(
+    ...     url="https://explorer.digitalearth.africa/stac/"
+    ... )
+    >>> search = catalog.search(
+    ...     bbox=[57.2, -20.6, 57.9, -19.9],  # xmin, ymin, xmax, ymax
+    ...     datetime=["2023-01-01T00:00:00Z", "2023-01-31T00:00:00Z"],
+    ...     collections=["s2_l2a"],
+    ... )
+    >>> dp = IterableWrapper(iterable=[search])
+    >>> dp_pystac_item_list = dp.list_pystac_items_by_search()
+    ...
+    >>> # Loop or iterate over the DataPipe stream
+    >>> it = iter(dp_pystac_item_list)
+    >>> stac_item = next(it)
+    >>> stac_item
+    <Item id=ec16dbf6-9729-5a8f-9d72-5e83a8b9f30d>
+    >>> stac_item.properties  # doctest: +NORMALIZE_WHITESPACE
+    {'title': 'S2B_MSIL2A_20230103T062449_N0509_R091_T40KED_20230103T075000',
+     'gsd': 10,
+     'proj:epsg': 32740,
+     'platform': 'sentinel-2b',
+     'view:off_nadir': 0,
+     'instruments': ['msi'],
+     'eo:cloud_cover': 0.02,
+     'odc:file_format': 'GeoTIFF',
+     'odc:region_code': '40KED',
+     'constellation': 'sentinel-2',
+     'sentinel:sequence': '0',
+     'sentinel:utm_zone': 40,
+     'sentinel:product_id': 'S2B_MSIL2A_20230103T062449_N0509_R091_T40KED_20230103T075000',
+     'sentinel:grid_square': 'ED',
+     'sentinel:data_coverage': 28.61,
+     'sentinel:latitude_band': 'K',
+     'created': '2023-01-03T06:24:53Z',
+     'sentinel:valid_cloud_cover': True,
+     'sentinel:boa_offset_applied': True,
+     'sentinel:processing_baseline': '05.09',
+     'proj:shape': [10980, 10980],
+     'proj:transform': [10.0, 0.0, 499980.0, 0.0, -10.0, 7900000.0, 0.0, 0.0, 1.0],
+     'datetime': '2023-01-03T06:24:53Z',
+     'cubedash:region_code': '40KED'}
+    """
+
+    def __init__(self, source_datapipe):
+        if pystac_client is None:
+            raise ModuleNotFoundError(
+                "Package `pystac_client` is required to be installed to use this datapipe. "
+                "Please use `pip install pystac-client` or "
+                "`conda install -c conda-forge pystac-client` "
+                "to install the package"
+            )
+        self.source_datapipe = source_datapipe
+
+    def __iter__(self):
+        for item_search in self.source_datapipe:
+            yield from item_search.items()
+
+    def __len__(self):
+        return sum(item_search.matched() for item_search in self.source_datapipe)
```

Comment:
Considered using this try-except statement in case `item_search.matched()` returns `None` and raises a `TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'`, but couldn't find a suitable STAC collection at https://radiantearth.github.io/stac-browser/ that doesn't implement `.matched()`. Leaving this out for now, but may add this back in if someone encounters the TypeError (and files a bug report that can be included in a regression test)!

```suggestion
        try:
            return sum(item_search.matched() for item_search in self.source_datapipe)
        except TypeError:  # unsupported operand type(s) for +: 'int' and 'NoneType'
            return sum(len(item_search.items()) for item_search in self.source_datapipe)
```